src/ParallelAlgorithms/SurfaceFinder/SurfaceFinder.cpp

```diff
@@ -0,0 +1,83 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ParallelAlgorithms/SurfaceFinder/SurfaceFinder.hpp"
+
+#include <array>
+#include <cstddef>
+#include <vector>
+
+#include "DataStructures/ApplyMatrices.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "Domain/Domain.hpp"
+#include "NumericalAlgorithms/RootFinding/TOMS748.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Gsl.hpp"
+
+// Wrapping for the interpolator for Toms748 rootfind.
+struct RayHelper {
+ public:
+  explicit RayHelper(const DataVector& ray)
+      : N_points_(ray.size()), values_(ray), extent_{ray.size()} {}
+  double operator()(double x) const {
+    std::array<Matrix, 1> int_arr{};
+    int_arr[0] =
+        Spectral::interpolation_matrix<Spectral::Basis::Legendre,
+                                       Spectral::Quadrature::GaussLobatto>(
+            N_points_, DataVector{1, x});
+    return apply_matrices(int_arr, values_, extent_)[0];
+  }
+
+ private:
+  size_t N_points_{};
+  DataVector values_{};
+  Index<1> extent_{};
+};
+
+namespace SurfaceFinder {
+
+std::vector<std::optional<double>> find_wedge_surface(
+    const double target, const Mesh<3>& mesh, const Scalar<DataVector>& data,
+    const tnsr::I<DataVector, 2, Frame::ElementLogical>& logical_direction,
+    const double tolerance) {
+  const size_t num_rays = logical_direction[0].size();
+  const DataVector subtracted_data = get(data) - target;
+  // The third Matrix in this array is not initialised so it is skipped by
+  // apply_matrices()
+  std::array<Matrix, 3> interp_array;
+  std::vector<std::optional<double>> result(num_rays, std::nullopt);
+  const double ray_size = mesh.extents(2);
+  DataVector interpolated_data(ray_size);
+  const auto xi_index = mesh.slice_through(0);
+  const auto eta_index = mesh.slice_through(1);
+
+  for (size_t i = 0; i < num_rays; i++) {
+    // Potential speed-up: Currently, the interpolation is done one ray at a
+    // time (i.e. the interpolation matrices are built to interpolate to one
+    // ray, and then applied to the data sequentially). Instead one could use
+    // something like irregular interpolant to do this all at once.
+    interp_array[0] =
+        Spectral::interpolation_matrix(xi_index, get<0>(logical_direction)[i]);
+    interp_array[1] =
+        Spectral::interpolation_matrix(eta_index, get<1>(logical_direction)[i]);
+
+    // Interpolate data onto the ray.
+    apply_matrices(make_not_null(&interpolated_data), interp_array,
+                   subtracted_data, mesh.extents());
+
+    // Perform root-find.
+    if (std::signbit(interpolated_data[0]) !=
+        std::signbit(interpolated_data[ray_size - 1])) {
```

Comment:
This doesn't work for Gauss quadrature because the outermost grid points are not at -1 and 1. Either disable for Gauss quadrature (emit an error if the radial mesh has Gauss quadrature) or just use the radial interpolant that you have already set up to evaluate the data at -1 and 1. For GaussLobatto quadrature you can still just take the outermost points directly.

Reply:
I think it shouldn't be disabled for Gauss quadrature, is it worth having a check for which quadrature it is or can I just use the interpolant always?

Reply:
It's an optimization 🤷‍♂️ if it's easy enough to do the switch between gauss and GaussLobatto then do it, else leave a comment.